### PR TITLE
Support static libraries within `.xcframework` binaries

### PR DESCRIPTION
**Pull request by luispadron**
## Summary

This is WIP effort to fully support static libraries `.a`. There a few edge cases we are not currently handling and this PR aims to correct them.

### Fixtures

This PR adds some new fixtures for testing XCFrameworks with static libraries

- `tests/ios/xcframeworks/StaticLib`
  - Contains the App target which imports all of the static library xcframeworks
  - Test with `bazel build //tests/ios/xcframeworks/StaticLib:StaticLibApp`
- `tests/ios/xcframeworks/ObjcLib`
  - Contains a fixture for a pure Objective-C static library in an XCFramework
  - Use the `tests/ios/xcframeworks/ObjcLib/MakeXCFramework.sh` script to update the XCFramework fixture if project changes

### Testing Objective-C failure without VFS :x:

```sh
bazel build //tests/ios/xcframeworks/StaticLib:StaticLibApp --cpu=arm64
INFO: Analyzed target //tests/ios/xcframeworks/StaticLib:StaticLibApp (3 packages loaded, 40 targets configured).
INFO: Found 1 target...
ERROR: /Users/lpadron/Development/rules_ios/tests/ios/xcframeworks/StaticLib/BUILD.bazel:3:16: Compiling tests/ios/xcframeworks/StaticLib/App/main.m failed: (Aborted): wrapped_clang failed: error executing command
  (cd /private/var/tmp/_bazel_lpadron/7c5e30dd733c097d91026d4bca2f7997/sandbox/darwin-sandbox/209/execroot/build_bazel_rules_ios && \
  exec env - \
    APPLE_SDK_PLATFORM=iPhoneSimulator \
    APPLE_SDK_VERSION_OVERRIDE=15.2 \
    PATH=/Users/lpadron/Library/Caches/bazelisk/downloads/bazelbuild/bazel-5.0.0-darwin-arm64/bin:/Users/lpadron/.rvm/gems/ruby-2.7.4/bin:/Users/lpadron/.rvm/gems/ruby-2.7.4@global/bin:/Users/lpadron/.rvm/rubies/ruby-2.7.4/bin:/Users/lpadron/.pyenv/shims:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/local/bin:/usr/local/munki:/Library/Apple/usr/bin:/Users/lpadron/.cargo/bin:/Users/lpadron/go/bin:/opt/homebrew/opt/go/libexec/bin:/Users/lpadron/.rvm/bin \
    XCODE_VERSION_OVERRIDE=13.2.1.13C100 \
    ZERO_AR_DATE=1 \
  external/local_config_cc/wrapped_clang -arch x86_64 '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -fno-omit-frame-pointer -g 'DEBUG_PREFIX_MAP_PWD=.' -include rules/library/common.pch -Wshorten-64-to-32 -Wbool-conversion -Wconstant-conversion -Wduplicate-method-match -Wempty-body -Wenum-conversion -Wint-conversion -Wunreachable-code -Wmismatched-return-types -Wundeclared-selector -Wuninitialized -Wunused-function -Wunused-variable -iquote . -iquote bazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin -iquote bazel-out/ios-x86_64-min9.0-applebin_ios-ios_x86_64-dbg-ST-ea30d59265dc/bin -MD -MF bazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/tests/ios/xcframeworks/StaticLib/_objs/StaticLibApp_objc/arc/main.d -DOS_IOS -fno-autolink -isysroot __BAZEL_XCODE_SDKROOT__ -F__BAZEL_XCODE_SDKROOT__/System/Library/Frameworks -F__BAZEL_XCODE_DEVELOPER_DIR__/Platforms/iPhoneSimulator.platform/Developer/Library/Frameworks -fobjc-arc '-mios-simulator-version-min=10.0' -fexceptions -fasm-blocks '-fobjc-abi-version=2' -fobjc-legacy-dispatch -Ibazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/tests/ios/xcframeworks/StaticLib/StaticLibApp_private_hmap.hmap -Ibazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/tests/ios/xcframeworks/StaticLib/StaticLibApp_public_hmap.hmap -Ibazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/tests/ios/xcframeworks/StaticLib/StaticLibApp_private_angled_hmap.hmap -iquotebazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/tests/ios/xcframeworks/StaticLib/StaticLibApp_private_hmap.hmap -fmodules '-fmodule-name=StaticLibApp' -I. -index-store-path bazel-out/rules_ios_global_index_store.indexstore '-fmodules-cache-path=bazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/_objc_module_cache' -O0 '-DDEBUG=1' -fstack-protector -fstack-protector-all -g -c tests/ios/xcframeworks/StaticLib/App/main.m -o bazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/tests/ios/xcframeworks/StaticLib/_objs/StaticLibApp_objc/arc/main.o)
# Configuration: 01e20dec9cb3563b1606cbfd9dca336219a362383e7ccebc192785e25f1af2d7
# Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox
tests/ios/xcframeworks/StaticLib/App/main.m:3:9: fatal error: 'ObjcLib/Foo.h' file not found
#import <ObjcLib/Foo.h>
        ^~~~~~~~~~~~~~~
1 error generated.
```

### Testing Objective-C failure with VFS ❌ 

```sh
bazel build //tests/ios/xcframeworks/StaticLib:StaticLibApp --features=apple.virtualize_frameworks --cpu=arm64
INFO: Analyzed target //tests/ios/xcframeworks/StaticLib:StaticLibApp (1 packages loaded, 18 targets configured).
INFO: Found 1 target...
ERROR: /Users/lpadron/Development/rules_ios/tests/ios/xcframeworks/StaticLib/BUILD.bazel:3:16: Compiling tests/ios/xcframeworks/StaticLib/App/main.m failed: (Aborted): wrapped_clang failed: error executing command
  (cd /private/var/tmp/_bazel_lpadron/7c5e30dd733c097d91026d4bca2f7997/sandbox/darwin-sandbox/216/execroot/build_bazel_rules_ios && \
  exec env - \
    APPLE_SDK_PLATFORM=iPhoneSimulator \
    APPLE_SDK_VERSION_OVERRIDE=15.2 \
    PATH=/Users/lpadron/Library/Caches/bazelisk/downloads/bazelbuild/bazel-5.0.0-darwin-arm64/bin:/Users/lpadron/.rvm/gems/ruby-2.7.4/bin:/Users/lpadron/.rvm/gems/ruby-2.7.4@global/bin:/Users/lpadron/.rvm/rubies/ruby-2.7.4/bin:/Users/lpadron/.pyenv/shims:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/local/bin:/usr/local/munki:/Library/Apple/usr/bin:/Users/lpadron/.cargo/bin:/Users/lpadron/go/bin:/opt/homebrew/opt/go/libexec/bin:/Users/lpadron/.rvm/bin \
    XCODE_VERSION_OVERRIDE=13.2.1.13C100 \
    ZERO_AR_DATE=1 \
  external/local_config_cc/wrapped_clang -arch x86_64 '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -fno-omit-frame-pointer -g 'DEBUG_PREFIX_MAP_PWD=.' -include rules/library/common.pch -Wshorten-64-to-32 -Wbool-conversion -Wconstant-conversion -Wduplicate-method-match -Wempty-body -Wenum-conversion -Wint-conversion -Wunreachable-code -Wmismatched-return-types -Wundeclared-selector -Wuninitialized -Wunused-function -Wunused-variable -iquote . -iquote bazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin -MD -MF bazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/tests/ios/xcframeworks/StaticLib/_objs/StaticLibApp_objc/arc/main.d -DOS_IOS -fno-autolink -isysroot __BAZEL_XCODE_SDKROOT__ -F__BAZEL_XCODE_SDKROOT__/System/Library/Frameworks -F__BAZEL_XCODE_DEVELOPER_DIR__/Platforms/iPhoneSimulator.platform/Developer/Library/Frameworks -fobjc-arc '-mios-simulator-version-min=10.0' -fexceptions -fasm-blocks '-fobjc-abi-version=2' -fobjc-legacy-dispatch -ivfsoverlaybazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/tests/ios/xcframeworks/StaticLib/StaticLibApp_vfs.yaml -F/build_bazel_rules_ios/frameworks -Ibazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/tests/ios/xcframeworks/StaticLib/StaticLibApp_private_hmap.hmap -Ibazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/tests/ios/xcframeworks/StaticLib/StaticLibApp_public_hmap.hmap -Ibazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/tests/ios/xcframeworks/StaticLib/StaticLibApp_private_angled_hmap.hmap -iquotebazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/tests/ios/xcframeworks/StaticLib/StaticLibApp_private_hmap.hmap -fmodules '-fmodule-name=StaticLibApp' -I. -index-store-path bazel-out/rules_ios_global_index_store.indexstore '-fmodules-cache-path=bazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/_objc_module_cache' -O0 '-DDEBUG=1' -fstack-protector -fstack-protector-all -g -c tests/ios/xcframeworks/StaticLib/App/main.m -o bazel-out/ios-x86_64-min10.0-applebin_ios-ios_x86_64-dbg-ST-0f1b0425081f/bin/tests/ios/xcframeworks/StaticLib/_objs/StaticLibApp_objc/arc/main.o)
# Configuration: 03a7eac56b2943601d9fa4181b4a629dcd6812a137d4943b1166c2556bb3bdf2
# Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox
/private/var/tmp/_bazel_lpadron/7c5e30dd733c097d91026d4bca2f7997/sandbox/darwin-sandbox/216/execroot/build_bazel_rules_ios/tests/ios/xcframeworks/StaticLib/App/main.m:2:9: fatal error: 'ObjcLib/Foo.h' file not found
#import <ObjcLib/Foo.h>
        ^~~~~~~~~~~~~~~
1 error generated.
```

### Tasks

- [X] Create reproducible test case for missing headers when importing via Objective-C
- [x] Support exporting public headers when consumed by Objective-C targets
- [x] Test support for both Objective-C with VFS enabled and disabled

### CODE REVIEW

File: rules/library.bzl [hunk 1]
```diff
@@ -684,17 +684,6 @@ def apple_library(name, library_tools = {}, export_private_headers = True, names
...
-            ["**/*.h"],
-        )
-        import_module_maps = native.glob(
-            ["**/*.modulemap"],
-        )
-        if len(import_module_maps) > 0:
-            import_module_map = import_module_maps[0]
-        else:
-            import_module_map = None
-        vfs_imported_framework = _find_imported_framework_name(import_headers)
```
Comment (luispadron) [inline rules/library.bzl:688]: None of this was used for the XCFramework case, most of the logic is handled in the `_xcframework` function.

**Comment by luispadron** [comment]
@jerrymarino Looking for your thoughts on how to fix the missing headers when VFS is enabled. I've narrowed down the issue to it being caused by an empty header map that gets created [here](https://github.com/bazel-ios/rules_ios/blob/master/rules/library.bzl#L780-L786). 

In the Static Library XCFramework case, the headers are defined within the `.xcframework/arm-64/Headers` directory. However, it looks like for the header map to be created `objc_hdrs` needs to have some data, which it only has if the `public_headers` property is defined on the `apple_library` rule. I've tried to define it as like:

```
apple_framework(
    name = "ObjcLib",
    module_name = "ObjcLib",
    platforms = {"ios": "9.0"},
    public_headers = glob(["ObjcLib.xcframework/**/Headers/**/*.h"]),
    vendored_xcframeworks = [
      ...
    ]
    ....
)
```

This allows us to build and import the headers as expected when VFS is enabled, but then it breaks when VFS is disabled due to duplicate imports. Since each XCFramework slice defines the headers:

```
tests/ios/xcframeworks/StaticLib/ObjcLib/ObjcLib.xcframework
├── Info.plist
├── ios-arm64
│   ├── Headers
│   │   └── ObjcLib
│   │       └── Foo.h
│   └── libObjcLib.a
└── ios-arm64_x86_64-simulator
    ├── Headers
    │   └── ObjcLib
    │       └── Foo.h
    └── libObjcLib.a
```

It fails with this error when building without VFS:

```
Error in fail: 
                [Error] Multiple files with the same name exists.

                See below for the list of paths found for each basename:

                {"Foo.h": ("tests/ios/xcframeworks/StaticLib/ObjcLib/ObjcLib.xcframework/ios-arm64/Headers/ObjcLib/Foo.h", "tests/ios/xcframeworks/StaticLib/ObjcLib/ObjcLib.xcframework/ios-arm64_x86_64-simulator/Headers/ObjcLib/Foo.h")}
```

I'm not sure how to proceed here, it looks like each slice should define the headers themselves, possibly as part of the `vendored_xcframeworks` dict and then we could select just the headers for the current slice we're building. Thoughts?

**Comment by luispadron** [comment]
Looking at it some more and it seems if we can get the appropriate headers for the static library to get passed into `objc_library` here: https://github.com/bazel-ios/rules_ios/blob/master/rules/library.bzl#L1004

Then the imports work as expected with VFS enabled, still struggling to get this working properly since I can't seem to understand how to get just the header slice from the correct `.xcframework` we need. Using a glob like `ObcLib.xcframework/**/Headers/**/*.h` will include headers from all the slices which leads to duplicate headers and a failure (when not in VFS).

**Comment by luispadron** [comment]
I've moved the Swift support to a separate PR to keep this one focused on Objective-C: #540

File: rules/library.bzl [hunk 2]
```diff
@@ -280,34 +336,6 @@ def _xcframework_slice(*, xcframework_name, identifier, platform, platform_varia
...
-    vfs_framework_name = imported_framework_name if imported_framework_name else xcframework_name
-    framework_vfs_overlay(
-        name = resolved_target_name_vfs_overlay,
-        framework_name = vfs_framework_name,
-        modulemap = import_module_map,
-        swiftmodules = import_swiftmodules,
-        hdrs = import_headers,
-        tags = _MANUAL,
-        extra_search_paths = xcframework_name,
-    )
```
Comment (luispadron) [inline rules/library.bzl:310]: Moving this logic to the individual branches, there are some differences between each type of framework and it made more sense to do that in each individual branch.

File: rules/library.bzl [hunk 3]
```diff
@@ -412,6 +444,21 @@ def _find_imported_framework_name(outputs):
...
+    Note: for a static library there is no .framework.
+    Instead, the .xcframework is the name of the static library.
+    """
+    for output in outputs:
+        if not ".xcframework" in output:
+            continue
+        prefix = output.split(".xcframework/")[0]
+        fw_name = prefix.split("/")[-1]
+        return fw_name
+    return None
```
Comment (luispadron) [inline rules/library.bzl:460]: Without this function, for static libraries there is no `.framework` so the existing `_find_imported_framework_name` always returned `None` which leads to the wrong name for the "framework" in the VFS

File: rules/library.bzl [hunk 4]
```diff
@@ -675,17 +722,6 @@ def apple_library(name, library_tools = {}, export_private_headers = True, names
...
-            ["**/*.h"],
-        )
-        import_module_maps = native.glob(
-            ["**/*.modulemap"],
-        )
-        if len(import_module_maps) > 0:
-            import_module_map = import_module_maps[0]
-        else:
-            import_module_map = None
-        vfs_imported_framework = _find_imported_framework_name(import_headers)
```
Comment (luispadron) [inline rules/library.bzl:688]: This code doesn't do anything since `vfs_imported_framework` isn't used here, instead this is created for each slice in the `_xcframework_slice` function.

File: rules/library.bzl [hunk 5]
```diff
@@ -267,10 +296,37 @@ def _xcframework_slice(*, xcframework_name, identifier, platform, platform_varia
...
     elif (linkage, packaging) == ("static", "library"):
+        # For a static library the `path` is the path to the library (`.a`) itself.
+        # We need the path to the top-level slice directory so we can find headers, modulemaps and swiftmodules.
+        slice_dir = paths.dirname(path)
+        import_headers, import_module_map, import_swiftmodules = _xcframework_slice_imports(slice_dir)
         native.objc_import(
             name = resolved_target_name,
             archives = [path],
+            hdrs = import_headers,
+            includes = ["%s/Headers" % slice_dir],
```
Comment (luispadron) [inline rules/library.bzl:319]: Similar to #543 we need to provide `hdrs` and `includes` to support `#import <...>`

File: rules/library.bzl [hunk 6]
```diff
@@ -267,10 +296,37 @@ def _xcframework_slice(*, xcframework_name, identifier, platform, platform_varia
...
+        imported_framework_name = _find_imported_xcframework_name(import_headers)
+        vfs_framework_name = imported_framework_name if imported_framework_name else xcframework_name
+        framework_vfs_overlay(
+            name = resolved_target_name_vfs_overlay,
+            framework_name = vfs_framework_name,
+            modulemap = import_module_map,
+            swiftmodules = import_swiftmodules,
+            hdrs = import_headers,
             tags = _MANUAL,
         )
```
Comment (luispadron) [inline rules/library.bzl:331]: The only changes here are:
 - The `framework_name` is now correct when using the new `_find_imported_xcframework_name` 
 - I removed `extra_search_paths` to copy the way we create the VFS overlay for normal static libraries. If we pass this parameter then importing fails for VFS, removing it fixes the issue

**Comment by luispadron** [comment]
@jerrymarino @jszumski This is ready for a proper review now that it works for both VFS and non-VFS modes 🎉

File: rules/library.bzl [hunk 7]
```diff
@@ -242,11 +242,28 @@ def _canonicalize_swift_version(swift_version):
 def _xcframework_build_type(*, linkage, packaging):
     return (linkage, packaging)
 
+def _xcframework_slice_imports(path):
+    """
+    Returns a tuple of the headers, module map and swiftmodules for the given path in a .xcframework.
+    """
+    import_headers = native.glob(["%s/**/*.h" % path])
+    import_module_maps = native.glob(["%s/**/*.modulemap" % path])
+    import_swiftmodules = native.glob(["%s/**/*.swiftmodule/*.*" % path])
```
Comment (jszumski) [inline rules/library.bzl:251]: Can you add `allow_empty = True` to these 3 globs to match https://github.com/bazel-ios/rules_ios/pull/518? Unfortunately we can't add a test case for this config since analysis tests can't transition on `incompatible_` flags.

Comment (luispadron) [reply rules/library.bzl]: Updated!

**Review by jszumski — APPROVED** [review]
Looks great!

File: rules/library.bzl (hunk 3 above)
Comment (jerrymarino) [reply rules/library.bzl:469]: At one point someone tried to make a nested xcframework - that should also probably be an error 👁️ . There isn't a spec for `.xcframework` but it seemed wrong. I would recommend to make an explicit error if they try to feed it code that looks like that: e.g. if you split `output` - it has many parts

**Review by jerrymarino — APPROVED** [review]
@luispadron LGTM - thanks for the PR 🙇‍♂️

File: rules/library.bzl (hunk 4 above)
Comment (luispadron) [reply rules/library.bzl:688]: Makes sense, I added a `fail` if `slices` is empty!

File: rules/library.bzl [hunk 8]
```diff
@@ -257,7 +283,19 @@ def _xcframework_slice(*, xcframework_name, identifier, platform, platform_varia
             dsym_imports = dsym_imports,
             tags = _MANUAL,
         )
+        imported_framework_name = _find_imported_framework_name(import_headers)
+        vfs_framework_name = imported_framework_name if imported_framework_name else xcframework_name
+        framework_vfs_overlay(
```
[the remark this answers is not in the document]
Comment (luispadron) [reply rules/library.bzl]: Cleaned this up, good idea!
